[RFC] Make OP-TEE physically relocatable when S-EL2 SPMC is not enabled

seonghp opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

This patch series try to dissociate `CFG_CORE_PHYS_RELOCATABLE` from `CFG_CORE_SEL2_SPMC`.

First, this patch allows OP-TEE to register ` [TZDRAM_BASE, TZDRAM_BASE + TZDRAM_SIZE)` as secure memory (i.e., `secure_only[0]`) when it cannot get secure memory information via FF-A manifest.

Then, this patch also tries to make use of memory regions below the physical offset. With certain scenarios such as physical ASLR, the physical offset can be large and it's not desirable to ignore that region.

Lastly, to make memory monitor remain honest, this patch introduce a mechanism to invalidate certain region within the memory pool. In particular, it invalidates overlapping region between the TEE RAM and the TA RAM so that this region is excluded from memory usage statistics.

Known remaining issue is that there remains redundant mappings, i.e., TA RAM mapping from some VA to PA that belongs to overlapped regions b/w TEE RAM and TA RAM, which I plan to address soon.

Looking forward to feedback from the community.

jenswi-linaro commented:
What is the use case for this?

seonghp commented:
Hi @jenswi-linaro, thank you for your response. In particular, we are interested in applying physical ASLR of the OP-TEE.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1618,6 +1618,8 @@ void __weak boot_save_args(unsigned long a0, unsigned long a1,
 #endif
 		}
+		if (IS_ENABLED(CFG_CORE_PHYS_RELOCATABLE))
+			core_mmu_set_secure_memory(TZDRAM_BASE, TZDRAM_SIZE);
```
Please use `TRUSTED_DRAM_BASE` and `TRUSTED_DRAM_SIZE` instead.

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -146,7 +146,7 @@ void core_mmu_get_ta_range(paddr_t *base, size_t *size)
 	static_assert(ARRAY_SIZE(secure_only) <= 2);
 	if (ARRAY_SIZE(secure_only) == 1) {
-		vaddr_t load_offs = 0;
+		vaddr_t load_offs __maybe_unused = 0;
```
Why is this `__maybe_unused`?

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -153,9 +153,14 @@ void core_mmu_get_ta_range(paddr_t *base, size_t *size)
 
 		assert(secure_only[0].size >
 		       load_offs + TEE_RAM_VA_SIZE + TEE_SDP_TEST_MEM_SIZE);
+#ifdef CFG_CORE_SEL2_SPMC
```
``` C
if (IS_ENABLED(CFG_CORE_SEL2_SPMC)) {
...
} else {
...
}
```

jenswi-linaro reviewed (commented):
That may require quite a bit of physical memory to be meaningful.

Anyway, I'm not so keen on "core: invalidate common RAM regions b/w TA and TEE", we should find a way to have a split up TA memory range.

seonghp commented:
> Anyway, I'm not so keen on "core: invalidate common RAM regions b/w TA and TEE", we should find a way to have a split up TA memory range.

Hi @jenswi-linaro, I'd love to hear your thoughts on my approach before I start coding.

The reason why a memory pool can only support up to one single contiguous memory regions is, to my understanding, because `tee_mm_pool_t` has one `lo` and `size`.

I think, by changing `tee_mm_pool_t` definition to:

```c
#define NUM_REGIONS 2

struct _tee_mm_pool_t {
	tee_mm_entry_t *entry;
	paddr_t lo[NUM_REGIONS]; /* low boundary of the pool */
	paddr_size_t size[NUM_REGIONS]; /* pool size */
	uint32_t flags; /* Config flags for the pool */
	uint8_t shift; /* size shift */
	unsigned int lock;
#ifdef CFG_WITH_STATS
	size_t max_allocated;
#endif
};
```
we can make the pool consists of split TA RAM regions.

Could you share your thoughts on this approach?

jenswi-linaro commented:
Adding
``` C
	paddr_t lo[NUM_REGIONS]; /* low boundary of the pool */
	paddr_size_t size[NUM_REGIONS]; /* pool size */
```
isn't enough since `tee_mm_get_smem()` needs to know the `lo` applicable for the entry. Also, this would allocate space for all the other pools that only need one range.

Unless I'm mistaken, multiple ranges are only needed for allocations from `tee_mm_sec_ddr`. Perhaps it's enough to do all those allocations via a helper function that knows to look in `tee_mm_sec_ddr2` or such if `tee_mm_sec_ddr` is full? Another option is to keep the `tee_mm_alloc2()` you added and accept that `max_allocated` isn't accurate.

In the slightly longer term, I think we should look into managing memory used by OP-TEE core more efficiently. Currently, we take `TEE_RAM_VA_SIZE` amount of physical memory, but we don't use all of it. It would be nice with a more dynamic approach where the unused physical memory could be reused. This is out of the scope of this PR though.

seonghp commented:
Thank you for your insights. I think having errors in `tee_mm_sec_ddr.max_allocated` is acceptable, and more preferable than adding extra logic. I revised the patch.

Another remaining issue is that currently there exist redundant mappings to TEE RAM physical memory.

```
              VA                 PA

    TA_RAM +------+---------->+------+
           |      |           |      |
           |      |           |      |
           +------+-------+-->+------+ core_mmu_
           |xxxxxx|      /    |      | tee_load_pa
overlapped |xxxxxx|     /     |      |
   area    |xxxxxx|    /      |      |
           |xxxxxx|   |       |      |
           +------+---+----+->+------+
           |      |   |   /   |      |
           |      |   /  /    |      |
           +------+  /  |     +------+
                    /   |
                   /    |
   TEE_RAM +------+    /
           |  RX  |   /
           +------+  /
           |  RW  | /
           |      |/
           +------+
```

jenswi-linaro commented:
Yes, the redundant mapping is an issue.

seonghp commented:
I decided to split TA RAM regions in the memory map, so there's no redundant mappings now.

```c
		if (IS_ENABLED(CFG_CORE_PHYS_RELOCATABLE) &&
		    !IS_ENABLED(CFG_CORE_SEL2_SPMC)) {
			ADD_PHYS_MEM(MEM_AREA_TA_RAM, ta_base,
				     core_mmu_tee_load_pa - ta_base);

			ADD_PHYS_MEM(MEM_AREA_TA_RAM,
				     core_mmu_tee_load_pa + TEE_RAM_VA_SIZE,
				     ta_size -
				     (core_mmu_tee_load_pa - ta_base) -
				     TEE_RAM_VA_SIZE);
		} else {
```

Also, I decided to add `tee_mm_sec_ddr_extra` for second TA RAM region and made a helper function that attempts to allocate from `tee_mm_sec_ddr_extra` when allocation from `tee_mm_sec_ddr` fails, and substituted it for `tee_mm_alloc()` calls.

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -1005,17 +1019,23 @@ static size_t collect_mem_ranges(struct tee_mmap_region *memory_map,
 {
 	const struct core_mmu_phys_mem *mem = NULL;
-	vaddr_t ram_start = secure_only[0].paddr;
+	vaddr_t tee_ram_start;
```
`= 0`

seonghp replied:
applied

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -1053,7 +1073,31 @@ static size_t collect_mem_ranges(struct tee_mmap_region *memory_map,
+		 * middle of the secure DRAM, rather than the secure DRAM base.
+		 * As a result, there may be up to two remaining memory regions
+		 * (i.e., one in the front and one in the back), which can be
+		 * used TA RAM.
```
"... can be used **for** TA RAM."

seonghp replied:
Thanks! applied

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -1350,13 +1394,19 @@ static unsigned long init_mem_map(struct tee_mmap_region *memory_map,
 	vaddr_t id_map_start = (vaddr_t)__identity_map_init_start;
 	vaddr_t id_map_end = (vaddr_t)__identity_map_init_end;
-	vaddr_t start_addr = secure_only[0].paddr;
+	vaddr_t tee_start_addr;
```
`= 0`

seonghp replied:
applied

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -2552,6 +2602,29 @@ static TEE_Result teecore_init_pub_ram(void)
+
+	for (map = get_memory_map(); !core_mmap_is_end_of_table(map); map++) {
+		if (map->type == MEM_AREA_TA_RAM) {
+			DMSG("map = { .va=%lx, .size=%lx }", map->va, map->size);
```
`			DMSG("map = { .va=%#"PRIxVA", .size=%#zx }", map->va, map->size);`

seonghp replied:
Oh this was used while debugging and I think this can be removed form the commit

etienne-lms commented on core/mm/tee_mm.c:
```diff
@@ -225,6 +225,18 @@ tee_mm_entry_t *tee_mm_alloc(tee_mm_pool_t *pool, size_t size)
+	mm = tee_mm_alloc(&tee_mm_sec_ddr, size);
+	if (!mm) {
+		return tee_mm_alloc(&tee_mm_sec_ddr_extra, size);
+	}
```
coding style: remove braces

etienne-lms reviewed (commented):
Some nitpicking comments.

jforissier commented:
How can this be tested? I'd like to verify this on QEMUv8 and perhaps add a line to the CI. To avoid increasing the CI time we could piggyback on make -j$(nproc) check CFG_CRYPTO_WITH_CE82=y` (see [ci.yml](https://github.com/OP-TEE/optee_os/blob/4.1.0/.github/workflows/ci.yml#L299)) and simply add the proper flags to that line.

seonghp commented:
Thanks @etienne-lms for the review. I believe I've addressed all your comments.

Thanks @jforissier for the feedback. To validate this patch, the recommended approach would be to cherry-pick seonghp/arm-trusted-firmware@f846124039 to TF-A and build OP-TEE with CFG_CORE_PHYS_RELOCATABLE=y.

However, enabling a CI for this change presents a challenge, as it relies on BL2 to load the OP-TEE image at a random physical offset, which falls outside the scope of this repository.

Contributing code to TF-A might be a potential solution, though it's not certain at this point.

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -2586,4 +2657,29 @@ void core_mmu_init_ta_ram(void)
+	if (IS_ENABLED(CFG_CORE_PHYS_RELOCATABLE) &&
+	    !IS_ENABLED(CFG_CORE_SEL2_SPMC)) {
+		if (!tee_mm_is_empty(&tee_mm_sec_ddr_extra))
+			panic("Extra TA RAM pool is not empty");
```
This "can't happen", an assert() should be enough.

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -2586,4 +2657,29 @@ void core_mmu_init_ta_ram(void)
+			panic("Extra TA RAM is not secure");
+
+		/* remove previous config and init Extra TA ddr memory pool */
+		tee_mm_final(&tee_mm_sec_ddr_extra);
```
`tee_mm_final()` isn't needed since the pool is empty.

jenswi-linaro commented on core/mm/tee_mm.c:
```diff
@@ -225,6 +225,17 @@ tee_mm_entry_t *tee_mm_alloc(tee_mm_pool_t *pool, size_t size)
+	tee_mm_entry_t *mm = NULL;
+
+	mm = tee_mm_alloc(&tee_mm_sec_ddr, size);
+	if (!mm)
```
I'd prefer the more direct logic:
``` C
if (mm)
     return mm;

return tee_mm_alloc(&tee_mm_sec_ddr_extra, size);
```

seonghp commented:
Thank you, @jforissier, for your suggestion. I will investigate the approach you suggested.

Also, thanks to @jenswi-linaro for the review and feedback. I was not familiar with how OP-TEE operates alongside an SPMC at S-EL2, so I was cautious and tried to avoid disrupting the code executed when `CFG_CORE_SPMC_AT_EL2` config is enabled.

I will take some time to clean up the logic, especially for the `SPMC_AT_EL=2` case, and will soon submit a revised patch.

gowthamithiagu commented:
As CFG_CORE_PHYS_RELOCATABLE is tightly coupled with CFG_CORE_SEL2_SPMC, 
couldn't separate that out with minor changes.

There is another use case where OP-TEE needs to be loaded and executed in a memory which is determined at run time. 
Pre assigned secure memory could be less in certain scenarios and OP-TEE needs more memory for its execution.

@seonghp
How is the allocation between TEE_RAM and TA_RAM done here?
Can you explain how the OP-TEE comes to know where it is going to be executed at run time?
Are you passing the OP-TEE load address from the lower level bootloaders?
